lib/wallet/cachedtxdb.js

```diff
@@ -0,0 +1,242 @@
+// @ts-check
+
+/*!
+ * cachedtxdb.js - persistent transaction pool with in-memory cache
+ * Copyright (c) 2021, Namebase Inc (MIT Licence).
+ * https://github.com/namebasehq/hsd
+ */
+
+'use strict';
+
+const assert = require('bsert');
+
+const TXDB = require('./txdb');
+const layout = require('./layout').txdb;
+const Outpoint = require('../primitives/outpoint');
+
+/** @typedef {import('./walletdb')} WalletDB */
+/** @typedef {import('../primitives/coin')} Coin */
+/** @typedef {import('./txdb').Credit} Credit */
+/** @typedef {import('./wallet')} Wallet')} */
+/** @typedef {import('./path')} Path} */
+
+/**
+ * @template K, V 
+ * @param {Map<K, V>} parent 
+ * @param {K} key 
+ * @param {() => V} makeNew
+ * @returns {V}
+ */
+ function getOrCreate(parent, key, makeNew) {
+  if(parent.has(key)) {
+    return parent.get(key);
+  }
+
+  const newVal = makeNew();
+  parent.set(key, newVal);
+
+  return newVal;
+}
+
+/**
+ * @template K1, K2, V
+ * @param {Map<K1, Map<K2, V>>} parent 
+ * @param {K1} key 
+ * @returns {Map<K2,V>}
+ */
+function ensureMap(parent, key) {
+  return getOrCreate(parent, key, () => new Map());
+}
+
+/**
+ * Duplicate a credit via encode/decode
+ * 
+ * @param {Credit} c 
+ * @returns {Credit}
+ */
+function dupCredit(c) {
+  c.encode();
+  /** @type {Credit} */
+  const dup = (/** @type {any} */ (TXDB.Credit.decode(c)));
+  dup.coin.hash = Buffer.from(c.coin.hash);
+  dup.coin.index = c.coin.index;
+  return dup;
+}
+
+class CachedTXDB extends TXDB {
+  /**
+   * Create a TXDB with in-memory cache for coin selection
+   * @param {WalletDB} wdb 
+   * @param {string | number} wid 
+   */
+  constructor(wdb, wid) {
+    super(wdb, wid);
+
+    /** @type {Map<string, Map<number, Credit>>} */
+    this.cachedCredits = new Map();
+    /** @type {Map<number, Map<string, Set<number>>>} */
+    this.cachedCoinsByAddress = new Map();
+  }
+
+  async _loadCredits() {
+    /** @type {Credit[]} */
+    const allCredits = await super.getCredits(-1);
+
+    for(const c of allCredits) {
+      this._insertCredit(c, await this.getPath(c.coin))
+    }
+  }
+
+  /**
+   * Inserta Credit into the cache
+   * @param {Credit} credit
+   * @param {Path} path
+   */
+  _insertCredit(credit, path) {
+    const {coin} = credit;
+    const hashHex = coin.hash.toString('hex');
+    ensureMap(this.cachedCredits, hashHex).set(coin.index, credit);
+    getOrCreate(
+      ensureMap(this.cachedCoinsByAddress, path.account),
+      hashHex,
+      () => new Set()).add(coin.index);
+  }
+
+  /**
+   * 
+   * @param {Wallet} wallet 
+   * @returns {Promise<any>}
+   */
+  async open(wallet) {
+    const result = await super.open(wallet);
+    await this._loadCredits();
```

Comment:
We might want to add locks to the cachedtxdb if we expect `this._loadCredits` to take a non negligible (seconds) amount of time. I'm wondering if super.open setting bucket to non null etc will cause other code to assume the txdb is ready, and call get outpoints etc before this._loadCredits` can complete. Execution might switch over because it's async here so it could happen

Reply:
I've added locking around calls to open() in the wallet code (the only place that uses it).

Reply:
I had to remove these. For some reason it would prevent the process from stopping cleanly in response to  Ctrl+C